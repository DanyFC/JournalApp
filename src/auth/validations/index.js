export { default as loginValidation } from './loginValidations'
export { default as registerValidation } from './registerValidation'