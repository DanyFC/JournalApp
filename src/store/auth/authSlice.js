import { createSlice } from '@reduxjs/toolkit'

export const authSlice = createSlice({
  name: 'auth',
  initialState: {
    displayName: null,
    email: null,
    errorMessage: null,
    photoURL: null,
    status: 'checking', // INFO: 'checking', 'not-authenticated', 'authenticated'
    uid: null
  },
  reducers: {
    checkCredential: (state) => {
      state.status = 'checking'
    },
    login: (state, { payload }) => {
      state.displayName = payload.displayName
      state.email = payload.email
      state.errorMessage = null
      state.photoURL = payload.photoURL
      state.status = 'authenticated'
      state.uid = payload.uid
    },
    logout: (state, { payload }) => {
      state.displayName = null
      state.email = null
      state.errorMessage = payload ? payload : null
      state.photoURL = null
      state.status = 'not-authenticated'
      state.uid = null
    }
  },
})

export const { checkCredential, login, logout } = authSlice.actions