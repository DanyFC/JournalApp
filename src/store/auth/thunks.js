import { checkCredential, login, logout } from './'
import { loginWithEmailPassword, logoutFirebase, registerUserWithEmailPassword, signInWithGoogle } from '../../firebase/providers'
import { clearState } from '../journal'

export const startLoginWithEmailPassword = (user) => {
  return async (dispatch) => {
    dispatch(checkCredential())
    const result = await loginWithEmailPassword(user)
    if (!result.ok) {
      dispatch(logout(result.errorMessage))
      return
    }
    dispatch(login(result))
  }
}

export const startGoogleSignIn = () => {
  return async (dispatch) => {
    dispatch(checkCredential())
    const result = await signInWithGoogle()
    if (!result.ok) {
      dispatch(logout(result.errorMessage))
      return
    }
    dispatch(login(result))
  }
}

export const startCreatingUserWithEmailPassword = (newUser) => {
  return async (dispatch) => {
    dispatch(checkCredential())
    const result = await registerUserWithEmailPassword(newUser)
    if (!result.ok) {
      dispatch(logout(result.errorMessage))
      return
    }
    dispatch(login(result))
  }
}

export const startLogout = () => {
  return async (dispatch) => {
    await logoutFirebase()
    dispatch(clearState())
    dispatch(logout())
  }
}