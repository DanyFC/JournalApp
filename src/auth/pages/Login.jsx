import { Alert, Button, Grid, Link, TextField, Typography } from '@mui/material'
import { Google } from '@mui/icons-material'
import { Link as RouterLink } from 'react-router-dom'
import { loginValidation } from './../validations'
import { startGoogleSignIn, startLoginWithEmailPassword } from './../../store/auth'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks'
import { useMemo } from 'react'
import Auth from '../layout/Auth'

const formData = {
  email: '',
  password: ''
}

const Login = () => {
  const dispatch = useDispatch()
  const { status, errorMessage } = useSelector(state => state.auth)

  const isAuthenticating = useMemo(() => status === 'checking', [status])

  const { email, password, formState, errors, onInputChange, onHandleBlur, onSubmit } = useForm(formData, loginValidation, signInWhitEmailPassword)

  function signInWhitEmailPassword() {
    dispatch(startLoginWithEmailPassword(formState))
  }

  const onGoogleSignIn = () => {
    dispatch(startGoogleSignIn())
  }

  return (
    <Auth title='Login'>
      <form onSubmit={onSubmit} aria-label='login-form'>
        <Grid container>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              fullWidth
              error={!!errors.email}
              helperText={errors.email}
              label='Email'
              name='email'
              onChange={onInputChange}
              onBlur={onHandleBlur}
              placeholder='example@domain.com'
              type='email'
              value={email}
            />
          </Grid>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              fullWidth
              error={!!errors.password}
              helperText={errors.password}
              inputProps={{ 'aria-label': 'Password' }}
              label='Password'
              name='password'
              onChange={onInputChange}
              onBlur={onHandleBlur}
              placeholder='●●●●●●●●'
              type='password'
              value={password}
            />
          </Grid>

          {errorMessage && (
            <Grid item xs={12} sx={{ mt: 2 }}>
              <Alert severity='error'>{errorMessage}</Alert>
            </Grid>
          )}

          <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
            <Grid item xs={12} sm={6}>
              <Button disabled={isAuthenticating} variant='contained' type='submit' fullWidth>
                Login
              </Button>
            </Grid>
            <Grid item xs={12} sm={6}>
              <Button disabled={isAuthenticating} variant='contained' fullWidth onClick={onGoogleSignIn} >
                <Google />
                <Typography sx={{ ml: 1 }}>Google</Typography>
              </Button>
            </Grid>
          </Grid>

        </Grid>

        <Grid container direction='row' justifyContent='end'>
          <Link component={RouterLink} color='inherit' to='/auth/register'>
            Create a new account!
          </Link>
        </Grid>

      </form>
    </Auth>


  )
}
export default Login