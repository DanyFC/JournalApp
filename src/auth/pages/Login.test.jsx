/**
 * @jest-environment jsdom
*/
import { authSlice } from '../../store/auth'
import { configureStore } from '@reduxjs/toolkit'
import { fireEvent, render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { notAuthenticatedState } from '../../fixtures/auth'
import { Provider } from 'react-redux'
import Login from './Login'

const store = configureStore({
  reducer: {
    auth: authSlice.reducer
  },
  preloadedState: {
    auth: notAuthenticatedState
  }
})

/* eslint-disable no-undef */
const mockStartGoogleSignIn = jest.fn()
const mockStartLoginWithEmailPassword = jest.fn()

jest.mock('../../store/auth/thunks', () => ({
  startGoogleSignIn: () => mockStartGoogleSignIn,
  startLoginWithEmailPassword: (formData) => () => mockStartLoginWithEmailPassword(formData)
}))

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: () => (fn) => fn()
}))

describe('Tests in <Login />', () => {

  beforeEach(() => jest.clearAllMocks())

  test.skip('Can be match with snapshot.', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  test.skip('Login with google must call startGoogleSignIn', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    )
    const googleButton = screen.getByRole('button', { name: 'Google' })

    fireEvent.click(googleButton)

    expect(mockStartGoogleSignIn).toHaveBeenCalled()
  })

  test('Submit must call startLoginWithEmailPassword with form values', () => {
    const formData = {
      email: 'user@gmail.com',
      password: '123456'
    }

    render(
      <Provider store={store}>
        <MemoryRouter>
          <Login />
        </MemoryRouter>
      </Provider>
    )
    const emailInput = screen.getByRole('textbox', { name: 'Email' })
    const passwordInput = screen.getByLabelText('Password')
    const form = screen.getByLabelText('login-form')

    fireEvent.change(emailInput, { target: { name: 'email', value: formData.email } })
    fireEvent.change(passwordInput, { target: { name: 'password', value: formData.password } })
    fireEvent.submit(form)

    expect(mockStartLoginWithEmailPassword).toHaveBeenCalledWith(formData)
  })
})
/* eslint-enable no-undef */
