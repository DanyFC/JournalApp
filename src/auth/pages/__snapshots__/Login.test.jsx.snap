// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <Login /> Can be match with snapshot. 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-ro1yt6-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3 box-shadow animate__animated animate__fadeInDown css-1ar3nvi-MuiGrid-root"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5 css-1jcm1rz-MuiTypography-root"
      >
        Login
      </h5>
      <form>
        <div
          class="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1hzi63y-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r0:"
                id=":r0:-label"
              >
                Email
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r0:"
                  name="email"
                  placeholder="example@domain.com"
                  type="email"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Email
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1hzi63y-MuiGrid-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for=":r1:"
                id=":r1:-label"
              >
                Password
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id=":r1:"
                  name="password"
                  placeholder="●●●●●●●●"
                  type="password"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Password
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-h80lhz-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
                disabled=""
                tabindex="-1"
                type="submit"
              >
                Login
              </button>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-1b3l6lk-MuiGrid-root"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-fullWidth css-1fu7jd5-MuiButtonBase-root-MuiButton-root"
                disabled=""
                tabindex="-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="GoogleIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12.545,10.239v3.821h5.445c-0.712,2.315-2.647,3.972-5.445,3.972c-3.332,0-6.033-2.701-6.033-6.032s2.701-6.032,6.033-6.032c1.498,0,2.866,0.549,3.921,1.453l2.814-2.814C17.503,2.988,15.139,2,12.545,2C7.021,2,2.543,6.477,2.543,12s4.478,10,10.002,10c8.396,0,10.249-7.85,9.426-11.748L12.545,10.239z"
                  />
                </svg>
                <p
                  class="MuiTypography-root MuiTypography-body1 css-yyvcc2-MuiTypography-root"
                >
                  Google
                </p>
              </button>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container css-1svu6jh-MuiGrid-root"
        >
          <a
            class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-xdxgn1-MuiTypography-root-MuiLink-root"
            href="/auth/register"
          >
            Create a new account!
          </a>
        </div>
      </form>
    </div>
  </div>
</div>
`;
